registry/storage/driver/walk.go

```diff
@@ -59,3 +59,41 @@ func WalkFallback(ctx context.Context, driver StorageDriver, from string, f Walk
 	}
 	return nil
 }
+
+// WalkFilesFallback traverses a filesystem defined within driver, starting
+// from the given path, calling f on each file. It uses the List method and Stat to drive itself.
+// If an error is returned from WalkFn, processing stops
+func WalkFilesFallback(ctx context.Context, driver StorageDriver, from string, f WalkFn) error {
+	children, err := driver.List(ctx, from)
+	if err != nil {
+		return err
+	}
+	sort.Stable(sort.StringSlice(children))
+	for _, child := range children {
+		// TODO(stevvooe): Calling driver.Stat for every entry is quite
+		// expensive when running against backends with a slow Stat
+		// implementation, such as s3. This is very likely a serious
+		// performance bottleneck.
+		fileInfo, err := driver.Stat(ctx, child)
+		if err != nil {
+			switch err.(type) {
```

Comment:
> WalkFiles does not have a special handling of ErrSkipDir to skip directory traversal without canceling the entire walk. Instead, any error will cancel the entire walk.

Why exactly didn't we add `special handling of ErrSkipDir to skip directory traversal without canceling the entire walk` _or_ why does it make sense not to?

Reply:
This impl is mostly just a copy-paste of `WalkFallback` with minor changes. To your point, it would be good to have another error type like `ErrStop` for canceling the entire walk, but the source of `ErrSkipDir` could come from a source outside of distribution (with how most WalkFn are implemented), so it could technically be a breaking change when this goes to upstream.